### PR TITLE
Feature/640 softcap improvements

### LINKED ISSUE #35: Rename starRate: starRatePer1000
So we dont confuse it with the previous regular `starRate`. Maybe also for the platform.

### PR DESCRIPTION
**Pull request by gorgos**
https://starbase.atlassian.net/browse/SP-640

Fixes #35

### CODE REVIEW

**Comment by gorgos** [comment]
I want to add unit-tests first, but putting this on hold for the TokenSale production deployments as well as automatic deployments feature.

**Comment by gorgos** [comment]
TODO:

- Forward funds immediately when softcap is set + softcap is reached. Maybe there's also a nice way to simply send all funds of the contract all at once. Then there's no need to call `forwardFunds()` manually.
- Unit tests

**Comment by gustavoguimaraes** [comment]
@gorgos This is finished :-)

**Comment by gorgos** [comment]
See https://starbaseteam.slack.com/archives/C87ANH3PH/p1548731303002900 for point 4.

**Comment by gustavoguimaraes** [comment]
@gorgos ready for another review

**Comment by gustavoguimaraes** [comment]
> 1. We absolutely need the STAR refund. Just imagine a TokenSale selling out, and the last person does a purchase of 3000 ETH, but may only receive tokens worth of 1 ETH.

I don't think it is a problem anymore. There is already a check for it on the buyTokens function
```
// remainder logic
if (tokensSold.add(tokens) > crowdsaleCap) {
     tokens = crowdsaleCap.sub(tokensSold);

      starAllocationToTokenSale = tokens.div(starRatePer1000).div(1000);
}
```

The TokenSale here won't take more than what the crowdsaleCap is. No more no less. No need to worry about the starRefund. I thought too much about it and added the `starRefund` variable.

> 2. After the Slack discussion, I think we can go with my idea from `4)`. Then simply add this: 'Therefore, we will only need a function so users could get their invested ETH or STAR back.' So a withdraw function of funds that checks if a soft cap was not reached and the sale has finished, both the case? -> send money back to requesting user.
Ok. I can add this.

> 3. Please [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) up the tests, many duplicate tests for the soft cap in there.
Can do this too.

**Comment by gorgos** [comment]
> I don't think it is a problem anymore. There is already a check for it on the buyTokens function

right, perfect. I'll add a withdraw function.

**Comment by gorgos** [comment]
@gustavoguimaraes Ive added the withdraw function, please review.

And I started with drying up the tests, but Ive only done it for one so far. Have a look at https://github.com/StarbaseCo/starbase-platform-smart-contracts/pull/30/commits/e867ec77052495fb2621d64c269fb1ce4c044ccf as reference.

Still missing:

(1) sells tokens up to crowdsale cap + refund in Wei
(2) sells tokens up to crowdsale cap + refund in Star
(3) sells tokens up to crowdsale cap (maybe can be removed completely? since we have it in 1 and 2)
(4) ends crowdsale when all tokens are sold
(5) ends crowdsale when all tokens are sold with wei

File: contracts/TokenSale.sol [hunk 1]
```diff
@@ -310,6 +316,23 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
...
+     * @dev withdraw funds for failed sales
+     */
+    function withdrawUserFunds() public {
+        require(hasEnded(), "Can only withdraw funds for ended sales!");
+        require(
+            !hasReachedSoftCap(),
+            "Can only withdraw funds for sales that didn't reach soft cap!"
+        );
+
+        uint256 investedEthRefund = ethInvestments[msg.sender];
```
Comment (gustavoguimaraes) [inline contracts/TokenSale.sol]: There is a reentrancy attack here. You should set the ethInvestments value back to zero

**Comment by gorgos** [comment]
> There is a reentrancy attack here. You should set the ethInvestments value back to zero

@gustavoguimaraes great catch! i will add a unit test for this

**Pull request merged**